xcdat/temporal.py

```diff
@@ -1079,49 +1139,90 @@ def _preprocess_dataset(self, ds: xr.Dataset) -> xr.Dataset:
 
         return ds
 
-    def _drop_incomplete_djf(self, dataset: xr.Dataset) -> xr.Dataset:
-        """Drops incomplete DJF seasons within a continuous time series.
-
-        This method assumes that the time series is continuous and removes the
-        leading and trailing incomplete seasons (e.g., the first January and
-        February of a time series that are not complete, because the December of
-        the previous year is missing). This method does not account for or
-        remove missing time steps anywhere else.
+    def _subset_coords_for_custom_seasons(
+        self, ds: xr.Dataset, months: List[str]
+    ) -> xr.Dataset:
+        """Subsets time coordinates to the months included in custom seasons.
 
         Parameters
         ----------
-        dataset : xr.Dataset
-            The dataset with some possibly incomplete DJF seasons.
+        ds : xr.Dataset
+            The dataset.
+        months : List[str]
+            A list of months included in custom seasons.
+            Example: ["Nov", "Dec", "Jan"]
 
         Returns
         -------
         xr.Dataset
-            The dataset with only complete DJF seasons.
+            The dataset with time coordinate subsetted to months used in
+            custom seasons.
         """
-        # Separate the dataset into two datasets, one with and one without
-        # the time dimension. This is necessary because the xarray .where()
-        # method concatenates the time dimension to non-time dimension data
-        # vars, which is not a desired behavior.
-        ds = dataset.copy()
-        ds_time = ds.get([v for v in ds.data_vars if self.dim in ds[v].dims])  # type: ignore
-        ds_no_time = ds.get([v for v in ds.data_vars if self.dim not in ds[v].dims])  # type: ignore
-
-        start_year, end_year = (
-            ds[self.dim].dt.year.values[0],
-            ds[self.dim].dt.year.values[-1],
-        )
-        incomplete_seasons = (f"{start_year}-01", f"{start_year}-02", f"{end_year}-12")
+        month_ints = sorted([MONTH_STR_TO_INT[month] for month in months])
 
-        for year_month in incomplete_seasons:
-            try:
-                coord_pt = ds.loc[dict(time=year_month)][self.dim][0]
-                ds_time = ds_time.where(ds_time[self.dim] != coord_pt, drop=True)
-            except (KeyError, IndexError):
-                continue
+        coords_by_month = ds.time.groupby(f"{self.dim}.month").groups
+        month_to_time_idx = {
+            k: coords_by_month[k] for k in month_ints if k in coords_by_month
+        }
+        month_to_time_idx = sorted(list(chain.from_iterable(month_to_time_idx.values())))  # type: ignore
```

Comment:
This chain of functions can be simplified by splitting it into separate variables.